dll/win32/browseui/regtreeoptions.cpp

```diff
@@ -1,59 +1,565 @@
+
+static HBITMAP CreatDibBitmap(HDC hDC, UINT cx, UINT cy, UINT bpp)
 {
+    BITMAPINFO bi;
+    LPVOID pvBits;
+
+    ZeroMemory(&bi, sizeof(bi));
+    bi.bmiHeader.biSize = sizeof(BITMAPINFOHEADER);
+    bi.bmiHeader.biWidth = cx;
+    bi.bmiHeader.biHeight = cy;
+    bi.bmiHeader.biPlanes = 1;
+    bi.bmiHeader.biBitCount = bpp;
+    bi.bmiHeader.biCompression = BI_RGB;
+    return CreateDIBSection(hDC, &bi, DIB_RGB_COLORS, &pvBits, NULL, 0);
 }
 
-CRegTreeOptions::~CRegTreeOptions()
+static HIMAGELIST CreateStateImages(UINT ImageSize)
 {
+    enum { bpp = 32 };
+    UINT Failed = FALSE, State;
+    HIMAGELIST hIL = ImageList_Create(ImageSize, ImageSize, ILC_MASK | bpp, 4, 0);
+    if (!hIL)
+        return hIL;
+
+    INT BorderSize = max(1, min(ImageSize / 16, 4));
+    RECT Rect;
+    HDC hDC = CreateCompatibleDC(NULL);
+
+    HBITMAP hbmMask = CreateBitmap(ImageSize, ImageSize, 1, 1, NULL);
+    Failed |= hbmMask == NULL;
+
+    HBITMAP hbmData = CreatDibBitmap(hDC, ImageSize, ImageSize, bpp);
+    Failed |= hbmData == NULL;
+
+    const HGDIOBJ hbmOld = SelectObject(hDC, hbmMask);
+    Failed |= hbmOld == NULL;
+    SetRect(&Rect, 0, 0, ImageSize, ImageSize);
+    FillRect(hDC, &Rect, HBRUSH(GetStockObject(WHITE_BRUSH)));
+    InflateRect(&Rect, -BorderSize, -BorderSize);
+    FillRect(hDC, &Rect, HBRUSH(GetStockObject(BLACK_BRUSH)));
+
+    SelectObject(hDC, hbmData);
+    State = DFCS_BUTTONCHECK | DFCS_FLAT | DFCS_MONO;
+    DrawFrameControl(hDC, &Rect, DFC_BUTTON, State | 000000000000 | 0000000000000);
+    Failed |= ImageList_Add(hIL, hbmData, hbmMask) < 0;
+    DrawFrameControl(hDC, &Rect, DFC_BUTTON, State | DFCS_CHECKED | 0000000000000);
+    Failed |= ImageList_Add(hIL, hbmData, hbmMask) < 0;
+    DrawFrameControl(hDC, &Rect, DFC_BUTTON, State | 000000000000 | DFCS_INACTIVE);
+    Failed |= ImageList_Add(hIL, hbmData, hbmMask) < 0;
+    DrawFrameControl(hDC, &Rect, DFC_BUTTON, State | DFCS_CHECKED | DFCS_INACTIVE);
+    Failed |= ImageList_Add(hIL, hbmData, hbmMask) < 0;
+
+    SelectObject(hDC, hbmMask);
+    SetRect(&Rect, 0, 0, ImageSize, ImageSize);
+    FillRect(hDC, &Rect, HBRUSH(GetStockObject(WHITE_BRUSH)));
+    InflateRect(&Rect, -BorderSize, -BorderSize);
+    DrawFrameControl(hDC, &Rect, DFC_BUTTON, DFCS_BUTTONRADIOMASK | DFCS_FLAT | DFCS_MONO);
+
+    SelectObject(hDC, hbmData);
+    State = DFCS_BUTTONRADIOIMAGE | DFCS_FLAT | DFCS_MONO;
+    DrawFrameControl(hDC, &Rect, DFC_BUTTON, State | 000000000000 | 0000000000000);
+    Failed |= ImageList_Add(hIL, hbmData, hbmMask) < 0;
+    DrawFrameControl(hDC, &Rect, DFC_BUTTON, State | DFCS_CHECKED | 0000000000000);
+    Failed |= ImageList_Add(hIL, hbmData, hbmMask) < 0;
+    DrawFrameControl(hDC, &Rect, DFC_BUTTON, State | 000000000000 | DFCS_INACTIVE);
+    Failed |= ImageList_Add(hIL, hbmData, hbmMask) < 0;
+    DrawFrameControl(hDC, &Rect, DFC_BUTTON, State | DFCS_CHECKED | DFCS_INACTIVE);
+    Failed |= ImageList_Add(hIL, hbmData, hbmMask) < 0;
```

Comment:
I wonder: Could it be possible to get blank images for the slots where the ImageList_Add could have failed?

Reply:
You mean add a blank image if adding failed? No, you can't, all images are stored in one big bitmap, a blank image is the same size as other images.